mhs-reference-implementation/mhs/routing/tests/test_sds_handler.py

```diff
@@ -39,17 +41,75 @@
 
 class TestMHSAttributeLookupHandler(TestCase):
 
+    def setUp(self) -> None:
+        self.cache = dc.DictionaryCache()
+        self.mocked_client = mocks.mocked_sds_client()
+
     @async_test
     async def test_get_endpoint(self):
-        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client())
+        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), self.cache)
         attributes = await handler.retrieve_mhs_attributes(ODS_CODE, INTERACTION_ID)
         for key, value in expected_mhs_attributes.items():
             self.assertEqual(value, attributes[key])
 
-        # Assert exact number of attributes, minus the unique values
+        # Assert exact number of attributes
         self.assertEqual(len(attributes), len(expected_mhs_attributes))
 
     @async_test
     async def test_no_client(self):
         with self.assertRaises(ValueError):
-            sds.MHSAttributeLookupHandler(None)
+            sds.MHSAttributeLookupHandler(None, self.cache)
+
+    @async_test
+    async def test_no_cache(self):
+        with self.assertRaises(ValueError):
+            sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), None)
+
+    @async_test
+    async def test_value_added_to_cache(self):
+        handler = sds.MHSAttributeLookupHandler(mocks.mocked_sds_client(), self.cache)
+        handler.sds_client = mocks.mocked_sds_client()
+        future = asyncio.Future()
+        future.set_result(None)
+
+        handler.cache.add_cache_value = MagicMock()
+        handler.cache.retrieve_mhs_attributes_value = MagicMock()
+
+        handler.cache.retrieve_mhs_attributes_value.return_value = future
+        handler.cache.add_cache_value.return_value = future
+
+        result = await handler.retrieve_mhs_attributes(ODS_CODE, INTERACTION_ID)
+
+        handler.cache.add_cache_value.assert_called_with(ODS_CODE, INTERACTION_ID, result)
+
+    @async_test
+    async def test_sds_not_called_when_value_in_cache(self):
+        await self.cache.add_cache_value("check", "not", "added")
+        handler = sds.MHSAttributeLookupHandler(MagicMock(), self.cache)
+
+        result = await handler.retrieve_mhs_attributes("check", "not")
+
+        self.assertEqual(result, "added")
+        handler.sds_client.assert_not_called()
+
+    @async_test
+    async def test_retrieving_cache_value_doesnt_reset_ttl(self):
+
+        cache = dc.DictionaryCache(expiry_time=3)
+        client = MagicMock()
+        future = asyncio.Future()
+        future.set_result({'attributes': 'testData'})
+
+        client.get_mhs_details.return_value = future
+        handler = sds.MHSAttributeLookupHandler(client, cache)
+
+        await handler.cache.add_cache_value("check", "not", "added")  # value in cache for 3 seconds
+
+        time.sleep(2)
```

Comment:
Again, better to avoid sleeping in tests.

Reply:
Fixed as suggested